test/framework/deployment_helpers.go

```diff
@@ -105,13 +113,90 @@ type WatchDeploymentLogsInput struct {
 // logMetadata contains metadata about the logs.
 // The format is very similar to the one used by promtail.
 type logMetadata struct {
-	Job       string `json:"job"`
-	Namespace string `json:"namespace"`
-	App       string `json:"app"`
-	Pod       string `json:"pod"`
-	Container string `json:"container"`
-	NodeName  string `json:"node_name"`
-	Stream    string `json:"stream"`
+	Job       string            `json:"job"`
+	Namespace string            `json:"namespace"`
+	App       string            `json:"app"`
+	Pod       string            `json:"pod"`
+	Container string            `json:"container"`
+	NodeName  string            `json:"node_name"`
+	Stream    string            `json:"stream"`
+	Labels    map[string]string `json:"labels,omitempty"`
+}
+
+// WatchDeploymentLogsByLabelSelector streams logs for all containers for all pods belonging to a deployment on the basis of label. Each container's logs are streamed
+// in a separate goroutine so they can all be streamed concurrently. This only causes a test failure if there are errors
+// retrieving the deployment, its pods, or setting up a log file. If there is an error with the log streaming itself,
+// that does not cause the test to fail.
+func WatchDeploymentLogsByLabelSelector(ctx context.Context, input WatchDeploymentLogsBySelectorInput) {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for WatchControllerLogs")
+	Expect(input.ClientSet).NotTo(BeNil(), "input.ClientSet is required for WatchControllerLogs")
+	Expect(input.Labels).NotTo(BeNil(), "input.Selector is required for WatchControllerLogs")
+
+	deploymentList := &appsv1.DeploymentList{}
+	Eventually(func() error {
+		return input.GetLister.List(ctx, deploymentList, client.MatchingLabels(input.Labels))
+	}, retryableOperationTimeout, retryableOperationInterval).Should(Succeed(), "Failed to get deployment for labels")
+
+	for _, deployment := range deploymentList.Items {
+		pods := &corev1.PodList{}
+		Expect(input.GetLister.List(ctx, pods, client.InNamespace(deployment.Namespace), client.MatchingLabels(input.Labels))).To(Succeed(), "Failed to list Pods for deployment %s/%s", deployment.Namespace, deployment.Name)
```

Comment:
We either should 
* not even list the deployments and directly list pods
* or use `.spec.selector.matchLabels` instead of `input.Labels` here (this would match the function name :-) )

Reply:
Lots of this seems similar to what is happening at `WatchDeploymentLogs`, maybe we could extract the common parts to a private function?

Reply:
I did some refactoring kindly check, the naming of the functions can definitely be redone. Let me know what are your suggestions.